src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_CAPACITY = 10;
+
+    private int size;
+
+    private Object[] dataArray;
+
+    public ArrayList() {
+        this.dataArray = new Object[START_CAPACITY];
+        this.size = 0;
+    }
+
+    private boolean sizeCheck() {
+        return size == dataArray.length;
+    }
+
+    private void grow(int minCapacity) {
+        int oldCapacity = dataArray.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity < minCapacity) {
+            newCapacity = minCapacity;
+        }
+        dataArray = Arrays.copyOf(dataArray, newCapacity);
```

Comment:
Consider using `System.arraycopy()` instead of `Arrays.copyOf()` when growing the array to avoid initializing a new array with `null` values, which is done by `Arrays.copyOf()`. This can be more efficient, especially for large arrays.